File: frontend/src/api/visualization.ts
import apiClient from './index'

export interface VisualStep {
  step_no: number
  line: number
  description: string
  variables: Record<string, unknown>
  visual_state: Record<string, unknown>
}

export interface VisualizeResult {
  execution_id: string
  language: string
  steps: VisualStep[]
  total_steps: number
}

export function executeAndVisualize(code: string, language = 'python') {
  return apiClient.post<VisualizeResult>('/visualize/execute', {
    code,
    language,
  })
}

File: backend/app/schemas/visualization.py
"""Visualization schemas."""
from pydantic import BaseModel
from typing import Any


class VisualStep(BaseModel):
    """执行步骤快照."""
    step_no: int
    line: int  # 当前行号
    description: str
    variables: dict[str, Any]  # 变量状态
    visual_state: dict[str, Any]  # 可视化数据（节点/数组/树结构）


class VisualizeExecuteResponse(BaseModel):
    """代码执行可视化响应."""
    execution_id: str
    language: str
    steps: list[VisualStep]
    total_steps: int
